File: python/binarySearchTree/colourFlip.py
from node import Node

class ColourFlip:
    t1 = None
    t2 = None
    
    def newArray(self, n):
        for i in range(1, n+1):
            self.t1 = self.avlInsert(self.t1, i, 'w')
            self.t2 = self.avlInsert(self.t2, i, 'b')

    def colour(self, key):
        node = self.find(self.t1, key)
        
        return node.colour

    def flip(self, x):
        # print('--------------->enter flip: ', x)
        l1, r1 = self.split(self.t1, x)
        # print('l1: ', self.bfsPrint(l1, 'key'))
        # print('r2: ', self.bfsPrint(r1, 'key'))
        l2, r2 = self.split(self.t2, x)
        # print('l2: ', self.bfsPrint(l2, 'key'))
        # print('r2: ', self.bfsPrint(r2, 'key'))
        self.t1 = self.merge(l1, r2)
        self.t2 = self.merge(l2, r1)

    def split(self, node, x):
        if node is None:
            return (None, None)

        if x < node.key:
            r1, r2 = self.split(node.left, x)
            r3 = self.avlTreeMergeWithRoot(r2, node.right, node)
            return (r1, r3)
        elif x > node.key:
            r1, r2 = self.split(node.right, x)
            r3 = self.avlTreeMergeWithRoot(node.left, r1, node)
            return (r3, r2)
        else:
            r1 = node.left
            node.left = None
            
            if r1 is not None:
                r1.parent = None
            node.parent = None
            return (r1, node)

    def merge(self, r1, r2):
        root = self.findLargestElement(r1)
        # print('largest element in left: ', root.key)
        r1 = self.getRoot(self.delete(root))
        # print('r1 root: ', r1.key if r1 is not None else r1)
        return self.avlTreeMergeWithRoot(r1, r2, root)

    def findLargestElement(self, node):
        if node.right is not None:
            return self.findLargestElement(node.right)
        return node

    def delete(self, node):
        if node.right is None:
            p = node.parent
            l = node.left

            if l is not None:
                l.parent = p
            
            if p is not None:
                if p.left is not None and p.left.key == node.key:
                    p.left = l
                if p.right is not None and p.right.key == node.key:
                    p.right = l
                return p
            else:
                return l
        else:
            nextNode = self.next(node)
            node.key = nextNode.key

            p = nextNode.parent
            r = nextNode.right

            if p.left is not None and p.left.key == nextNode.key:
                p.left = r
            if p.right is not None and p.right.key == nextNode.key:
                p.right = r

            if r is not None:
                r.parent = p
            
            return p

    def getRoot(self, node):
        if node is None:
            return None

        if node.parent is None:
            return node
        return self.getRoot(node.parent)

    def next(self, node):
        if node.right is not None:
            return self.leftDescendant(node.right)
        else:
            return self.rightAncestor(node.parent, node.key)

    def leftDescendant(self, node):
        if node.left is not None:
            return self.leftDescendant(node.left)
        return node

    def rightAncestor(self, node, key):
        if key < node.key:
            return node
        
        if node.parent is not None:
            return self.rightAncestor(node.parent, key)
        return None

    def avlTreeMergeWithRoot(self, r1, r2, root):
        if abs(self.get(r1, 'height') - self.get(r2, 'height')) <= 1:
            self.mergeWithRoot(r1, r2, root)
            self.adjustHeightAndSize(root)
            
            return root
        elif self.get(r1, 'height') > self.get(r2, 'height'):
            r = self.avlTreeMergeWithRoot(r1.right, r2, root)
            r1.right = r
            r.parent = r1
            r1.parent = None

            return self.rebalance(r1)
        elif self.get(r1, 'height') < self.get(r2, 'height'):
            r = self.avlTreeMergeWithRoot(r1, r2.left, root)
            r2.left = r
            r.parent = r2
            r2.parent = None

            return self.rebalance(r2)

    def mergeWithRoot(self, r1, r2, root):
        root.left = r1
        root.right = r2
        root.parent = None

        if r1 is not None:
            r1.parent = root
        if r2 is not None:
            r2.parent = root

        return root

    def bfsPrint(self, root, key):
        result = []
        q = []
        q.append(root)

        while (len(q) > 0):
            tmp = q.pop(0)
            result.append(str(getattr(tmp, key)))

            if tmp.left is not None:
                q.append(tmp.left)

            if tmp.right is not None:
                q.append(tmp.right)

        return ','.join(result)

    def avlInsert(self, root, key, colour):
       root = self.insert(root, key, colour)
       node = self.find(root, key)
       
       return self.rebalance(node)

    def insert(self, node, key, colour):
        if node is None:
            return Node(key, None, 1, 1, colour)

        p = self.find(node, key)
        new = Node(key, p, 1, 1, colour)
        if key > p.key:
            p.right = new
        else:
            p.left = new

        return node
    
    def find(self, node, key):
        if key == node.key:
            return node
        
        if key > node.key:
            if node.right is None:
                return node
            return self.find(node.right, key)
        else:
            if node.left is None:
                return node
            return self.find(node.left, key)

    def rebalance(self, node):
        if self.get(node.left, 'height') > self.get(node.right, 'height') + 1:
            self.rebalanceRight(node)

        if self.get(node.right, 'height') > self.get(node.left, 'height') + 1:
            self.rebalanceLeft(node)

        self.adjustHeightAndSize(node)

        p = node.parent
        if p is not None:
            return self.rebalance(p)
        return node

    def get(self, node, key):
        return 0 if node is None else getattr(node, key)

    def rebalanceRight(self, node):
        l = node.left
        if self.get(l.right, 'height') > self.get(l.left, 'height'):
            self.rotateLeft(l)
        self.rotateRight(node)

    def rebalanceLeft(self, node):
        r = node.right
        if self.get(r.left, 'height') > self.get(r.right, 'height'):
            self.rotateRight(r)
        self.rotateLeft(node)

    def rotateRight(self, node):
        p = node.parent
        l = node.left
        r = l.right

        if p is not None:
            if p.left is not None and p.left.key == node.key:
                p.left = l
            if p.right is not None and p.right.key == node.key:
                p.right = l
        l.parent = p

        l.right = node
        node.parent = l

        node.left = r
        if r is not None:
            r.parent = node

        self.adjustHeightAndSize(node)
        self.adjustHeightAndSize(l)

    def rotateLeft(self, node):
        p = node.parent
        r = node.right
        l = r.left

        if p is not None:
            if p.left is not None and p.left.key == node.key:
                p.left = r
            if p.right is not None and p.right.key == node.key:
                p.right = r
        r.parent = p

        r.left = node
        node.parent = r

        node.right = l
        if l is not None:
            l.parent = node

        self.adjustHeightAndSize(node)
        self.adjustHeightAndSize(r)

    def adjustHeightAndSize(self, node):
        node.height = max(self.get(node.left, 'height'), self.get(node.right, 'height')) + 1
        node.size = self.get(node.left, 'size') + self.get(node.right, 'size') + 1

File: java/src/binarySearchTree/ColourFlip.java
package binarySearchTree;

import java.util.Queue;
import java.util.LinkedList;
import java.util.StringJoiner;
import java.util.Map;

public class ColourFlip {
    Node t1;
    Node t2;

    public String bfsPrint(Node node, String type) {
        Queue<Node> q = new LinkedList<Node>();
        StringJoiner sj = new StringJoiner(",");
        q.add(node);

        while(!q.isEmpty()) {
            Node tmp = q.poll();
            sj.add(tmp.getDataByType(type));

            if (tmp.left != null)
                q.add(tmp.left);

            if (tmp.right != null)
                q.add(tmp.right);
        }

        return sj.toString();
    }

    public void newArray(int n) {
        for (int i = 1 ; i < n + 1; i++) {
            this.t1 = this.avlInsert(this.t1, i, 'w');
            this.t2 = this.avlInsert(this.t2, i, 'b');
        }
    }

    public Node avlInsert(Node root, int key, char colour) {
        root = this.insert(root, key, colour);
        Node node = this.find(root, key);
        return this.rebalance(node);
    }

    private Node insert(Node root, int key, char colour) {
        if (root == null)
            return new Node(key, null, colour);

        Node node = this.find(root, key);
        Node newNode = new Node(key, node, colour);

        if (key < node.key)
            node.left = newNode;
        else
            node.right = newNode;

        return root;
    }

    private Node find(Node node, int key) {
        if (key == node.key)
            return node;

        if (key < node.key) {
            if (node.left == null)
                return node;
            return this.find(node.left, key);
        } else {
            if (node.right == null)
                return node;
            return this.find(node.right, key);
        }
    }

    private Node rebalance(Node node) {
        if (this.getHeight(node.left) > this.getHeight(node.right) + 1)
            this.rebalanceRight(node);

        if (this.getHeight(node.right) > this.getHeight(node.left) + 1)
            this.rebalanceLeft(node);

        this.adjustHeightAndSize(node);

        Node p = node.parent;
        if (p != null)
            return this.rebalance(p);
        else
            return node;
    }

    private int getHeight(Node node) {
        return node == null? 0 : node.height;
    }

    private int getSize(Node node) {
        return node == null ? 0 : node.size;
    }

    private void adjustHeightAndSize(Node node) {
        node.height = Math.max(this.getHeight(node.left), this.getHeight(node.right)) + 1;
        node.size = this.getSize(node.left) + this.getSize(node.right) + 1;
    }

    private void rebalanceRight(Node node) {
        Node l = node.left;
        if (this.getHeight(l.right) > this.getHeight(l.left))
            this.rotateLeft(l);
        this.rotateRight(node);
    }

    private void rebalanceLeft(Node node) {
        Node r = node.right;
        if (this.getHeight(r.left) > this.getHeight(r.right))
            this.rotateRight(r);
        this.rotateLeft(node);
    }

    private void rotateRight(Node node) {
        Node p = node.parent;
        Node l = node.left;
        Node r = l.right;

        if (p != null) {
            if (p.left != null && p.left.key == node.key)
                p.left = l;
            if (p.right != null && p.right.key == node.key)
                p.right = l;
        }
        l.parent = p;

        l.right = node;
        node.parent = l;

        node.left = r;
        if (r != null)
            r.parent = node;

        this.adjustHeightAndSize(node);
        this.adjustHeightAndSize(l);
    }

    private void rotateLeft(Node node) {
        Node p = node.parent;
        Node r = node.right;
        Node l = r.left;

        if (p != null) {
            if (p.left != null && p.left.key == node.key)
                p.left = r;
            if (p.right != null && p.right.key == node.key)
                p.right = r;
        }
        r.parent = p;

        r.left = node;
        node.parent = r;

        node.right = l;
        if (l != null)
            l.parent = node;

        this.adjustHeightAndSize(node);
        this.adjustHeightAndSize(r);
    }

    public Node merge(Node r1, Node r2) {
        Node root = this.findLargest(r1);
        r1 = this.avlDelete(root);
        //return mergeWithRoot(r1, r2, root);
        return avlMergeWithRoot(r1, r2, root);
    }

    private Node findLargest(Node node) {
        if (node.right != null)
            return this.findLargest(node.right);
        return node;
    }

    private Node avlDelete(Node node) {
        return this.rebalance(this.delete(node));
    }

    private Node delete(Node node) {
        if (node.right == null) {
            Node p = node.parent;
            Node l = node.left;

            if (l != null) {
                l.parent = p;
            }

            if (p != null) {
                if (p.left != null && p.left.key == node.key)
                    p.left = l;
                if (p.right != null && p.right.key == node.key)
                    p.right = l;

                return p;
            } else {
                return l;
            }
        } else {
            Node l = this.next(node);
            Node p = l.parent;
            Node r = l.right;

            node.key = l.key;
            p.left = r;
            if (r != null)
                r.parent = p;

            return p;
        }
    }

    private Node next(Node node) {
        if (node.right != null)
            return this.leftDescendant(node.right);
        else
            return this.rightAncestor(node.key, node.parent);
    }

    private Node leftDescendant(Node node) {
        if (node.left != null)
            return this.leftDescendant(node.left);
        else
            return node;
    }

    private Node rightAncestor(int key, Node node) {
        if (key == node.key)
            return node;

        if (node.parent != null)
            return this.rightAncestor(key, node.parent);
        else
            return null;
    }

    private Node mergeWithRoot(Node r1, Node r2, Node root) {
        root.left = r1;
        root.right = r2;

        if (r1 != null)
            r1.parent = root;
        if (r2 != null)
            r2.parent = root;

        return root;
    }

    private Node avlMergeWithRoot(Node r1, Node r2, Node root) {
        if (Math.abs(this.getHeight(r1) - this.getHeight(r2)) <= 1) {
            root = this.mergeWithRoot(r1, r2, root);
            this.adjustHeightAndSize(root);

            return root;
        } else if (this.getHeight(r1) > this.getHeight(r2)) {
            Node tmp = this.avlMergeWithRoot(r1.right, r2, root);
            r1.right = tmp;
            tmp.parent = r1;
            r1.parent = null;

            return this.rebalance(r1);
        } else if (this.getHeight(r2) > this.getHeight(r1)) {
            Node tmp = this.avlMergeWithRoot(r1, r2.left, root);
            r2.left = tmp;
            tmp.parent = r2;
            r2.parent = null;

            return this.rebalance(r2);
        }
        return null;
    }

    public Map<String, Node> split(Node node, int key) {
        return null;
    }

    public void flip(int idx) {

    }
}
